### PR TITLE
Fix development crasher on file deletion

### LINKED ISSUE #2542: Dev server crashes when files are deleted/renamed
Steps to reproduce:

```sh
gatsby develop
rm src/pages/mypage.js
```

Console output:
```
./node_modules/.bin/gatsby --version
1.1.10
yarn start v1.0.2
$ gatsby develop --port 4567
success delete html files from previous builds — 0.027 s
success open and validate gatsby-config.js — 0.008 s
success copy gatsby files — 0.022 s
success onPreBootstrap — 0.009 s
success source and transform nodes — 0.033 s
success building schema — 0.224 s
success createLayouts — 0.032 s
success createPages — 0.012 s
success createPagesStatefully — 0.036 s
success onPreExtractQueries — 0.002 s
success extract queries from components — 0.133 s
success run graphql queries — 0.021 s
success write out page data — 0.003 s
success write out redirect data — 0.001 s
success update schema — 0.091 s
success onPostBootstrap — 0.003 s

info bootstrap finished - 2.319 s

error UNHANDLED REJECTION


  Error: ENOENT: no such file or directory, open '/[redacted]/code/github.com/[redacted]/src/pages/[redacted].js'

  - Error

  - file-parser.js:239 FileParser._callee3$
    [...]/[gatsby]/dist/internal-plugins/query-runner/file-parser.js:239:25

  - file-parser.js:287 FileParser.parseFile
    [...]/[gatsby]/dist/internal-plugins/query-runner/file-parser.js:287:20

  - file-parser.js:304
    [...]/[gatsby]/dist/internal-plugins/query-runner/file-parser.js:304:30

  - Array.map

  - file-parser.js:303 FileParser._callee4$
    [...]/[gatsby]/dist/internal-plugins/query-runner/file-parser.js:303:67
```

Impact: Gatsby develop crashes during development.

If this issue is confirmed, I'd be happy to try to fix it. I think the fix will involve catching the ENOENT error, or preventing it in the first place by rechecking for file existence in file-parser.js.

Thanks!

### PR DESCRIPTION
**Pull request by mickeyreiss**
Posting 77405d7 as the changes were helpful for my investigation but should not be merged.

See 9bd26ac for a full explanation.

Fixes #2542.

cc/ @KyleAMathews

### CODE REVIEW

**Comment by mickeyreiss** [comment]
I was not sure what testing is typically done in this project. I verified that redux updates as expected and I am going to wait for CI to see whether any tests need to be updated.

File: packages/gatsby/src/internal-plugins/query-runner/file-parser.js [hunk 1]
```diff
@@ -117,8 +117,17 @@ const cache = {}
 
 export default class FileParser {
   async parseFile(file: string): Promise<?DocumentNode> {
-    // TODO figure out why fs-extra isn't returning a promise
-    const text = fs.readFileSync(file, `utf8`)
+    console.warn(`parsing file ${file}`);
+    let text;
+    try {
+      text = await fs.readFile(file, `utf8`)
```
Comment (mickeyreiss) [inline packages/gatsby/src/internal-plugins/query-runner/file-parser.js]: Although this change is part of the throwaway commit, I couldn't reproduce the issue mentioned in the TODO. Seems like the conversion to async io "just works."

File: packages/gatsby/src/internal-plugins/query-runner/query-compiler.js [hunk 2]
```diff
@@ -79,6 +79,7 @@ class Runner {
     // FIXME: this should all use gatsby's configuration to determine parsable
     // files (and how to parse them)
     let files = glob.sync(`${this.fragmentsDir}/**/*.+(t|j)s?(x)`)
+    console.warn(files)
```
Comment (KyleAMathews) [inline packages/gatsby/src/internal-plugins/query-runner/query-compiler.js]: These are for debugging right?

**Comment by KyleAMathews** [comment]
Played with this some more and super nice! Nice fix! Cleaning up edge bugs like this really are what makes the UX of a tool shine vs. feeling kinda crappy all the time.

Did a bit more work as well to fix returning to the development 404 page after the file was deleted as that wasn't handled correctly either.

Thanks for diving in and figuring things out and fixing the problem!

File: examples/using-remark/package.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
     "gatsby-remark-images": "latest",
     "gatsby-remark-katex": "latest",
     "gatsby-remark-prismjs": "latest",
-    "gatsby-remark-responsive-iframe": "latest",
+    "gatsby-remark-responsive-iframe": "^1.4.9",
```
Comment (mickeyreiss-visor) [inline examples/using-remark/package.json]: Was this change related to the issue, or is it something else?

Comment (KyleAMathews) [reply examples/using-remark/package.json]: No, I just ran yarn upgrade in this example site before using it to test your PR

**Comment by jlengstorf** [comment]
Hiya @mickeyreiss! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs

**Pull request merged**